### PR TITLE
feat: table widget label type columns

### LINKED ISSUE #4333: [Feature] Introduce a 'label' style for table columns
We need to introduce a way for users to show certain data(especially statuses) as a label. See screenshots below from Airtable and Internal.

Screen Shot 2021-05-05 at 9 45 47 PM

Screen Shot 2021-05-05 at 9 45 41 PM

Screen Shot 2021-05-05 at 9 47 26 PM

Properties to show:
Screen Shot 2021-05-05 at 9 48 39 PM

# Developer Handoff Document in Figma
Link to the developer Handoff Document: 
https://www.figma.com/file/dUpW1PrTAgZIboP8QeP9F6/Table-tags-Fri-25-June?node-id=2%3A1

### PR DESCRIPTION
**Pull request by vicky-primathon**
## Description
Label styles for column in table widget

Fixes #4333

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/table-widget-label-type-columns 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.87 **(-0.01)** | 36.8 **(-0.08)** | 34.25 **(0.02)** | 55.39 **(-0.01)**
 :red_circle: | app/client/src/components/ads/ColorPickerComponent.tsx | 37.84 **(0)** | 46.67 **(-16.97)** | 0 **(0)** | 40 **(0)**
 :sparkles: :new: | **app/client/src/components/designSystems/appsmith/TagComponent.tsx** | **64.71** | **14.29** | **16.67** | **60**
 :sparkles: :new: | **app/client/src/components/propertyControls/LabelColorPickerControl.tsx** | **40.54** | **14.29** | **18.18** | **40**
 :green_circle: | app/client/src/components/propertyControls/index.ts | 85.29 **(0.44)** | 100 **(0)** | 0 **(0)** | 84.85 **(0.47)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.47 **(-0.24)** | 38.26 **(-0.87)** | 36.21 **(0)** | 55.08 **(-0.27)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableStyledWrappers.tsx | 68.18 **(0.36)** | 21.53 **(0.57)** | 50 **(0)** | 68.18 **(0.36)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableUtilities.tsx | 36.97 **(-2.89)** | 19.51 **(-3.79)** | 13.95 **(-1.84)** | 35.57 **(-2.89)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/index.tsx | 26.6 **(-0.13)** | 5.94 **(-0.04)** | 26.42 **(0)** | 26.49 **(-0.14)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/propertyConfig.ts | 48.21 **(9.48)** | 10.2 **(0)** | 73.17 **(7.54)** | 49.69 **(9.54)**</details>

### CODE REVIEW

File: app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts [hunk 1]
```diff
@@ -154,6 +161,9 @@ export interface ColumnProperties {
   inputFormat?: string;
   dropdownOptions?: string;
   onOptionChange?: string;
+  tagSize?: string;
```
Comment (SatishGandham) [inline app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts:164]: Shouldn't these be combined into an object?

File: app/client/src/components/propertyControls/LabelColorPickerControl.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  "#9177FF",
+  "#FF76FE",
+  "#61DF48",
+  "#6698FF",
+  "#F8C356",
+  "#6C4CF1",
+  "#C5CD90",
+  "#6272C8",
+  "#4F70FD",
+  "#F0F0F0",
```
Comment (SatishGandham) [inline app/client/src/components/propertyControls/LabelColorPickerControl.tsx:23]: I think you forgot transparent here?

**Review by SatishGandham — CHANGES_REQUESTED** [review]

File: app/client/src/components/propertyControls/LabelColorPickerControl.tsx (hunk 2 above)
Comment (vicky-primathon) [reply app/client/src/components/propertyControls/LabelColorPickerControl.tsx:23]: @SatishGandham  Labels cannot be transparent

File: app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts (hunk 1 above)
Comment (vicky-primathon) [reply app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts:174]: @SatishGandham these are column level properties and not table level properties. Column properties are created based on these props. So if i combined these into an object, accessing these properties while rendering property pane will not be possible.

**Review by SatishGandham — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@vicky-primathon Observations so far:
1. Column type name is Label. As per the design it should be Tag
2. Other properties are not coming in property pane - I am not able to test other flows due to this.
3. After selecting Label for an column, the column data is coming as +3
Screenshot 2021-09-02 at 5 18 09 PM

**Comment by vicky-primathon** [comment]
> @vicky-primathon Observations so far:
> 
> 1. Column type name is Label. As per the design it should be Tag
> 2. Other properties are not coming in property pane - I am not able to test other flows due to this.
> 3. After selecting Label for an column, the column data is coming as +3
> 
> Screenshot 2021-09-02 at 5 18 09 PM

1. Issue is fixed
3. If the column size is smaller than we show the tag with tooltip. To see the tags, increase the size of column.

File: app/client/src/components/propertyControls/LabelColorPickerControl.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+> {
+  constructor(props: LabelColorPickerControlProps) {
+    super(props);
+    this.state = {
+      labelColors: {},
+    };
+  }
+
+  getLabels = () => {
+    const { tableData } = this.props.widgetProperties;
```
Comment (vicky-primathon) [inline app/client/src/components/propertyControls/LabelColorPickerControl.tsx:46]: @somangshu  this property is not accessible anymore

Comment (somangshu) [reply app/client/src/components/propertyControls/LabelColorPickerControl.tsx:46]: @riodeuno will need your help here with the feature design

**Comment by somangshu** [comment]
@vicky-primathon lets plan a meeting around this with the stakeholders so that we have what we need, Kindly let me know what help you might need here

**Comment by somangshu** [comment]
Closed this for now

**Pull request closed without merging**